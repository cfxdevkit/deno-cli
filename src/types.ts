export interface Block {
	blockNumber?: bigint
	timestamp: bigint | number
	number?: bigint // For espace block
}

export interface BlockInfo {
	number: bigint
	timestamp: string
	label: string
}

export interface Transaction {
	hash: string
	from: string
	to: `0x${string}` | null
	value: bigint | string
	gas: bigint
	gasPrice?: bigint | undefined
}

// --- Types ---
export type KeystoreEntry = {
	type: 'plaintext' | 'encoded'
	label: string
	mnemonic: string
}

export type KeystoreFile = {
	keystore: KeystoreEntry[]
	activeIndex: number | null
}
