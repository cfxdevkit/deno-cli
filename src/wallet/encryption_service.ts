import { crypto } from '@std/crypto'
import { Secret } from 'cliffy/prompt'

export class EncryptionService {
	async deriveKeyFromPassword(reason: string, salt: Uint8Array): Promise<CryptoKey> {
		const password = await Secret.prompt({ message: reason })
		const encoder = new TextEncoder()
		const passwordKey = await crypto.subtle.importKey(
			'raw',
			encoder.encode(password),
			'PBKDF2',
			false,
			['deriveKey'],
		)

		return crypto.subtle.deriveKey(
			{
				name: 'PBKDF2',
				salt,
				iterations: 100_000,
				hash: 'SHA-256',
			},
			passwordKey,
			{ name: 'AES-GCM', length: 256 },
			true,
			['encrypt', 'decrypt'],
		)
	}

	getRandomValues(array: Uint8Array) {
		return crypto.getRandomValues(array)
	}

	async encryptMnemonic(mnemonic: string): Promise<string> {
		const iv = this.getRandomValues(new Uint8Array(12))
		const salt = this.getRandomValues(new Uint8Array(16))
		const key = await this.deriveKeyFromPassword('Enter encryption password to secure your mnemonic.', salt)
		const encrypted = await crypto.subtle.encrypt(
			{ name: 'AES-GCM', iv },
			key,
			new TextEncoder().encode(mnemonic),
		)
		const encryptedData = new Uint8Array([...salt, ...iv, ...new Uint8Array(encrypted)])
		return btoa(String.fromCharCode(...encryptedData))
	}

	async decryptMnemonic(encryptedMnemonic: string): Promise<string> {
		const encryptedBytes = Uint8Array.from(atob(encryptedMnemonic), (c) => c.charCodeAt(0))
		const salt = encryptedBytes.slice(0, 16)
		const iv = encryptedBytes.slice(16, 28)
		const data = encryptedBytes.slice(28)

		for (let attempt = 1; attempt <= 3; attempt++) {
			try {
				const key = await this.deriveKeyFromPassword('Password: ', salt)
				const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, data)
				return new TextDecoder().decode(decrypted)
			} catch {
				console.error(`Decryption failed (${attempt}/3). Incorrect password or corrupted data.`)
			}
		}
		throw new Error('Maximum decryption attempts reached.')
	}
}
