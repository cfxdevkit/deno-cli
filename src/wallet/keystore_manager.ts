import { ensureFileSync } from '@std/fs'
import { join } from '@std/path'
import { KeystoreEntry, KeystoreFile } from '../types.ts'

export interface IKeystoreManager {
	readKeystore(): Promise<KeystoreFile | null>
	writeKeystore(): Promise<void>
	getKeystore(): KeystoreEntry[]
	setKeystore(keystore: KeystoreEntry[]): void
	getActiveIndex(): number
	setActiveIndex(index: number | null): void
}

export class KeystoreManager implements IKeystoreManager {
	public keystorePath: string
	public keystore: KeystoreEntry[] = []
	public activeIndex: number | null = 0

	constructor() {
		this.keystorePath = join(Deno.env.get('HOME') || '', '.devkit.keystore.json')
		ensureFileSync(this.keystorePath)
	}

	async readKeystore(): Promise<KeystoreFile | null> {
		try {
			const data = await Deno.readTextFile(this.keystorePath)
			return data.trim() ? JSON.parse(data) : null
		} catch (error) {
			if (error instanceof Deno.errors.NotFound) return null
			throw error
		}
	}

	async writeKeystore(): Promise<void> {
		const data: KeystoreFile = { keystore: this.keystore, activeIndex: this.activeIndex }
		await Deno.writeTextFile(this.keystorePath, JSON.stringify(data, null, 2))
	}

	getKeystore() {
		return this.keystore
	}

	setKeystore(keystore: KeystoreEntry[]) {
		this.keystore = keystore
	}

	setKeystorePath(keystore_path: string) {
		this.keystorePath = keystore_path
	}

	setActiveIndex(index: number | null) {
		this.activeIndex = index
	}

	getActiveIndex(): number {
		return this.activeIndex || 0
	}
}
