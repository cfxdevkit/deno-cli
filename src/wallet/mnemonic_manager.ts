import { Input, Select } from 'cliffy/prompt'
import { english } from 'viem/accounts'
import { KeystoreManager } from './keystore_manager.ts'
import { EncryptionService } from './encryption_service.ts'
import { HDWallet } from '@conflux-dev/hdwallet'

export class MnemonicManager {
	private keystoreManager: KeystoreManager
	public encryptionService: EncryptionService

	constructor(keystoreManager: KeystoreManager, encryptionService: EncryptionService) {
		this.keystoreManager = keystoreManager
		this.encryptionService = encryptionService
	}

	async addMnemonic(): Promise<void> {
		const storageChoice = await Select.prompt({
			message: 'Choose storage option for the mnemonic:',
			options: [
				{ name: 'Store encrypted', value: 'e' },
				{ name: 'Store in plaintext', value: 'p' },
			],
		})

		const mnemonic = await this.promptForMnemonic()

		const defaultLabel = `Mnemonic ${this.keystoreManager.getKeystore().length + 1}`
		const label = await Input.prompt({
			message: 'Enter a label for this mnemonic:',
			default: defaultLabel,
		})

		if (storageChoice === 'p') {
			this.keystoreManager.getKeystore().push({ type: 'plaintext', label, mnemonic })
		} else {
			const encryptedMnemonic = await this.encryptionService.encryptMnemonic(mnemonic)
			this.keystoreManager.getKeystore().push({ type: 'encoded', label, mnemonic: encryptedMnemonic })
		}

		await this.keystoreManager.writeKeystore()
		console.error(storageChoice === 'p' ? 'Mnemonic stored in plaintext.' : 'Mnemonic stored securely.')
	}

	private async promptForMnemonic(): Promise<string> {
		const userChoice = await Select.prompt({
			message: 'Generate or import a mnemonic?',
			options: [
				{ name: 'Generate a new mnemonic', value: 'g' },
				{ name: 'Insert an existing mnemonic', value: 'i' },
			],
		})
		return userChoice === 'i' ? await this.importMnemonic() : this.generateMnemonic()
	}

	generateMnemonic(): string {
		return HDWallet.generateMnemonic()
	}

	private async importMnemonic(): Promise<string> {
		console.log('\nPlease enter your mnemonic key one word at a time.')
		const words: string[] = []
		for (let i = 1; i <= 12; i++) {
			const word = await Input.prompt({
				message: `Enter word ${i} of 12`,
				suggestions: english,
				validate: (input) =>
					english.includes(input) || 'Invalid word. Please enter a valid BIP-39 mnemonic word.',
			})
			words.push(word)
		}
		return words.join(' ')
	}
}
