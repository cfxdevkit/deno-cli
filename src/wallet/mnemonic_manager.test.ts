import { assertEquals } from '@std/assert'
import { ansi } from 'cliffy/ansi'
import { snapshotTest } from 'cliffy/testing'
import { EncryptionService } from './encryption_service.ts'
import { MnemonicManager } from './mnemonic_manager.ts'
import { IKeystoreManager } from './keystore_manager.ts'
import { KeystoreEntry, KeystoreFile } from '../types.ts'

export class MockKeystoreManager implements IKeystoreManager {
	public keystorePath: string = '.mock_keystore.json' // Change to "protected"
	public keystore: KeystoreEntry[] = []
	public activeIndex: number | null = 0
	// deno-lint-ignore require-await
	async readKeystore(): Promise<KeystoreFile | null> {
		console.log('Mock readKeystore called.')
		return { keystore: this.keystore, activeIndex: this.activeIndex }
	}
	// deno-lint-ignore require-await
	async writeKeystore(): Promise<void> {
		console.log('Mock writeKeystore called.')
	}

	getKeystore(): KeystoreEntry[] {
		return this.keystore
	}

	setKeystore(keystore: KeystoreEntry[]): void {
		this.keystore = keystore
	}

	getActiveIndex(): number {
		return this.activeIndex || 0
	}

	setActiveIndex(index: number | null): void {
		this.activeIndex = index
	}

	setKeystorePath(keystorePath: string): void {
		this.keystorePath = keystorePath
	}
}

const mnemonic = 'test test test test test test test test test test test junk'
const mockKeystoreManager = new MockKeystoreManager()
const mockEncryptionService = new EncryptionService()
mockEncryptionService.getRandomValues = (typedArray: Uint8Array) => {
	return typedArray
}
// deno-lint-ignore require-await
mockEncryptionService.encryptMnemonic = async (_mnemonic) => `encrypted:${mnemonic}`

const mnemonicManager = new MnemonicManager(mockKeystoreManager, mockEncryptionService)
mnemonicManager.generateMnemonic = () => mnemonic

await snapshotTest({
	name: 'MnemonicManager - Add generated mnemonic with encryption',
	meta: import.meta,
	stdin: ansi
		.text('\n') // Select "Store encrypted"
		.text('\n') // Select "Generate a new mnemonic"
		.text('\n') // Select "accept default label"
		//   .text("\n") // Input "encryption password (mock don't needs password"
		.toArray(),
	async fn() {
		mockKeystoreManager.setKeystore([])
		mockKeystoreManager.setActiveIndex(0)

		await mnemonicManager.addMnemonic()

		const keystore = mockKeystoreManager.getKeystore()
		assertEquals(keystore.length, 1)
		assertEquals(keystore[0].type, 'encoded')
		assertEquals(keystore[0].mnemonic, `encrypted:${mnemonic}`)
		assertEquals(keystore[0].label, 'Mnemonic 1')
	},
})

await snapshotTest({
	name: 'MnemonicManager - Add generated mnemonic plaintext',
	meta: import.meta,
	stdin: ansi
		.cursorDown
		.text('\n') // Select "Store plaintext"
		.text('\n') // Select "Generate a new mnemonic"
		.text('\n') // Select "accept default label"
		.toArray(),
	async fn() {
		mockKeystoreManager.setKeystore([])
		mockKeystoreManager.setActiveIndex(0)

		await mnemonicManager.addMnemonic()

		const keystore = mockKeystoreManager.getKeystore()
		console.log(keystore)
		assertEquals(keystore.length, 1)
		assertEquals(keystore[0].type, 'plaintext')
		assertEquals(keystore[0].mnemonic.split(' ').length, 12)
		assertEquals(keystore[0].label, 'Mnemonic 1')
	},
})

// Test: Import mnemonic and store as plaintext
await snapshotTest({
	name: 'MnemonicManager - Import mnemonic plaintext',
	meta: import.meta,
	stdin: ansi
		.cursorDown
		.text('\n') // Select "Store plaintext"
		.cursorDown
		.text('\n') // Select "Insert an existing mnemonic"
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('junk\n')
		.text('\n') // Accept default label
		.toArray(),
	async fn() {
		mockKeystoreManager.setKeystore([])
		mockKeystoreManager.setActiveIndex(0)

		await mnemonicManager.addMnemonic()

		const keystore = mockKeystoreManager.getKeystore()
		assertEquals(keystore.length, 1)
		assertEquals(keystore[0].type, 'plaintext')
		assertEquals(keystore[0].mnemonic, mnemonic)
		assertEquals(keystore[0].label, 'Mnemonic 1')
	},
})

// Test: Import mnemonic and store as encrypted
await snapshotTest({
	name: 'MnemonicManager - Import mnemonic encrypted',
	meta: import.meta,
	stdin: ansi
		.text('\n') // Select "Store encrypted"
		.cursorDown
		.text('\n') // Select "Insert an existing mnemonic"
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('test\n')
		.text('junk\n')
		.text('\n') // Accept default label
		.toArray(),
	async fn() {
		mockKeystoreManager.setKeystore([])
		mockKeystoreManager.setActiveIndex(0)

		await mnemonicManager.addMnemonic()

		const keystore = mockKeystoreManager.getKeystore()
		assertEquals(keystore.length, 1)
		assertEquals(keystore[0].type, 'encoded')
		assertEquals(keystore[0].mnemonic, `encrypted:${mnemonic}`)
		assertEquals(keystore[0].label, 'Mnemonic 1')
	},
})

// // Test: Error handling for invalid mnemonic input
// await snapshotTest({
//     name: "MnemonicManager - Error on invalid mnemonic input",
//     meta: import.meta,
//     stdin: ansi
// 	.cursorDown
// 	.text("\n") // Select "Store plaintext"
// 	.cursorDown
// 	.text("\n") // Select "Insert an existing mnemonic"
// 	// .text("invalidword\n")
// 	// .eraseLine
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("test\n")
// 	.text("junk\n")
// 	.text("\n") // Accept default label
// 	.toArray(),
// async fn() {
//         mockKeystoreManager.setKeystore([]);
//         mockKeystoreManager.setActiveIndex(0);
// 		console.log(await mnemonicManager.addMnemonic())
//         // await assertRejects(() => mnemonicManager.addMnemonic(), Error, "Invalid word. Please enter a valid BIP-39 mnemonic word.");
//         assertEquals(mockKeystoreManager.getKeystore().length, 0);
//     },
// });

// Test: Multiple mnemonics stored
await snapshotTest({
	name: 'MnemonicManager - Multiple mnemonics stored',
	meta: import.meta,
	stdin: ansi
		.cursorDown
		.text('\n') // Select "Store plaintext"
		.text('\n') // Select "Generate a new mnemonic"
		.text('\n') // Select "accept default label"
		.text('\n') // Select "Store encrypted"
		.text('\n') // Select "Generate a new mnemonic"
		.text('\n') // Select "accept default label"
		.toArray(),
	async fn() {
		mockKeystoreManager.setKeystore([])
		mockKeystoreManager.setActiveIndex(0)

		await mnemonicManager.addMnemonic()
		await mnemonicManager.addMnemonic()

		const keystore = mockKeystoreManager.getKeystore()
		assertEquals(keystore.length, 2)
		assertEquals(keystore[0].type, 'plaintext')
		assertEquals(keystore[0].mnemonic, mnemonic)
		assertEquals(keystore[1].type, 'encoded')
		assertEquals(keystore[1].mnemonic, `encrypted:${mnemonic}`)
		assertEquals(keystore[0].label, 'Mnemonic 1')
		assertEquals(keystore[1].label, 'Mnemonic 2')
	},
})
