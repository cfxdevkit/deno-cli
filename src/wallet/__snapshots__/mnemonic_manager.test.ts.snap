export const snapshot = {};

snapshot[`MnemonicManager - Add generated mnemonic with encryption 1`] = `
stdout:
"? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store encrypted
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Generate a new mnemonic
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 1) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 1) › Mnemonic 1
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
"
stderr:
"Mnemonic stored securely.
"
`;

snapshot[`MnemonicManager - Add generated mnemonic plaintext 1`] = `
stdout:
'? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic:
  Store encrypted
❯ Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store in plaintext
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Generate a new mnemonic
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 1) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 1) › Mnemonic 1
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
[
  {
    type: "plaintext",
    label: "Mnemonic 1",
    mnemonic: "test test test test test test test test test test test junk"
  }
]
'
stderr:
"Mnemonic stored in plaintext.
"
`;

snapshot[`MnemonicManager - Import mnemonic plaintext 1`] = `
stdout:
"? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic:
  Store encrypted
❯ Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store in plaintext
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic?
  Generate a new mnemonic
❯ Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Insert an existing mnemonic
\\x1b[?25h\\x1b[?25h
Please enter your mnemonic key one word at a time.
? Enter word 1 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 1 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 2 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 2 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 3 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 3 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 4 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 4 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 5 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 5 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 6 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 6 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 7 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 7 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 8 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 8 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 9 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 9 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 10 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 10 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 11 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 11 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 12 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 12 of 12 › junk
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 1) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 1) › Mnemonic 1
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
"
stderr:
"Mnemonic stored in plaintext.
"
`;

snapshot[`MnemonicManager - Import mnemonic encrypted 1`] = `
stdout:
"? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store encrypted
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic?
  Generate a new mnemonic
❯ Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Insert an existing mnemonic
\\x1b[?25h\\x1b[?25h
Please enter your mnemonic key one word at a time.
? Enter word 1 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 1 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 2 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 2 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 3 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 3 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 4 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 4 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 5 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 5 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 6 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 6 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 7 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 7 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 8 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 8 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 9 of 12 › \\x1b[24G\\x1b[G\\x1b[0J
? Enter word 9 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 10 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 10 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 11 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 11 of 12 › test
\\x1b[?25h\\x1b[?25h? Enter word 12 of 12 › \\x1b[25G\\x1b[G\\x1b[0J
? Enter word 12 of 12 › junk
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 1) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 1) › Mnemonic 1
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
"
stderr:
"Mnemonic stored securely.
"
`;

snapshot[`MnemonicManager - Multiple mnemonics stored 1`] = `
stdout:
"? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic:
  Store encrypted
❯ Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store in plaintext
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Generate a new mnemonic
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 1) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 1) › Mnemonic 1
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
? Choose storage option for the mnemonic:
❯ Store encrypted
  Store in plaintext\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Choose storage option for the mnemonic: › Store encrypted
\\x1b[?25h\\x1b[?25h? Generate or import a mnemonic?
❯ Generate a new mnemonic
  Insert an existing mnemonic\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Generate or import a mnemonic? › Generate a new mnemonic
\\x1b[?25h\\x1b[?25h? Enter a label for this mnemonic: (Mnemonic 2) › \\x1b[51G\\x1b[G\\x1b[0J
? Enter a label for this mnemonic: (Mnemonic 2) › Mnemonic 2
\\x1b[?25h\\x1b[?25hMock writeKeystore called.
"
stderr:
"Mnemonic stored in plaintext.
Mnemonic stored securely.
"
`;
